### DISCUSSION TITLE
Probabilistic programming

### DISCUSSION THREAD

**Discussion by lukego** [kaveh808/kons-9#152] [Show and tell]
I'm taking my first baby-steps with using Kons-9 to visualize probabilistic programming results. It's fun!

Here's my very first plot that shows some random samples next to their underlying distribution, a slightly awkward test case of a Gaussian with a notch cut out on one side:

![Screenshot from 2022-10-26 10-27-24](https://user-images.githubusercontent.com/13791/197979772-e43692c1-aac4-477a-bf9f-aa176f41dc1c.png)

Even at this basic level it's already more fun for me to use Kons-9 than Blender, Julia, or R.

The next thing I need to figure out is how to draw a line/curve through a series of points and how to change colors :grinning: 

Context/code: https://twitter.com/lukego/status/1585187363474485249

Thanks @kaveh808 for this already great tool!

**Comment by kaveh808** [comment]
Good stuff. Check out `curve.lisp`. You can set point colors, but not sure if curves render them. Let me know if you have questions.

**Comment by lukego** [comment]
Hey cool! kons-9 is helping me to debug a Sequential Monte Carlo (particle filter) simulation today :grin: 

The simulation calls kons-9 to plot its internal state in a new color after each step. This way I can watch the simulation proceeding in real time. Being able to pan/scale interactively also helps for making sense of information without endlessly tweaking plotting settings.

This is actually really helpful. The convergence is happening much more suddenly than I expected (from right to left below) and this gives me an idea about where to look for a bug that's causing bad results in other higher-dimensional simulations.

Thanks @kaveh808 for a great tool! Too bad you weren't at ELS2023 this week because @foretspaisibles and I would have loved to buy you a beer :beers: 

![image](https://user-images.githubusercontent.com/13791/234835541-b9621c98-a7d7-4be0-83d2-6da4e9508176.png)

**Reply by kaveh808** [reply]
Hi @lukego ,

Wonderful! This made my day. SO glad you're finding the system useful.

I just watched the videos from ELS on Twitch. Wish I could have been there. Someone suggested that I make a demonstration proposal for next year.

P.S. you can set the bg to black by doing `(set-theme-dark)`

**Reply by aykaramba** [reply]
I love seeing people use Kons9 in such interesting ways.  Post more, would love to watch / read.

**Reply by lukego** [reply]
Thanks for the warm feedback! I posted some more notes on this little side-quest here: https://github.com/lukego/blog/issues/37

**Comment by kayomarz** [comment]
This is really inspiring @lukego! I've been playing around with kons-9 since a couple of weeks and it is indeed awesome thanks to @kaveh808!

btw @lukego I like your way of using git issues for blogging and might do just the same:)

**Comment by lukego** [comment]
Here is another Monte Carlo diagnostic that takes better advantage of 3D space: https://github.com/lukego/blog/issues/38

**Reply by kaveh808** [reply]
Very cool. If you think of any new kons-9 features that might help your work, please let me know.

**Reply by lukego** [reply]
> Very cool. If you think of any new kons-9 features that might help your work, please let me know.

The things that come to mind immediately are using an orthographic projection and being able to rescale the axes independently in a convenient way.

Orthographic projection would let me to take a head-on/side-on/top-down look at all the layers at the same time.

Rescaling the axes (e.g. fit-to-scene and interactive drag-to-rescale) would let me relax about visualizing abstract data without a common scale. Maybe my data has X values [-1000,1000] and Y values [-inf, 0] and Z-values [0.00000001,1]. To see this data in 3D requires some normalization and editorial decisions like where to clip outliers and it might make more sense to do that in the viewer (kons-9) than the data model (lisp code to rewrite the values with made-up units.)

I also need a better way to understand what I'm looking at e.g. to click or mouse-over a point in a pointcloud and get some information about it.

That's just a raw braindump to capture some thoughts that might otherwise disappear.

**Reply by kaveh808** [reply]
Eminently sensible requests for data visualization.

**Reply by lukego** [reply]
This might be starting to overlap with the domain of [Paraview](https://www.paraview.org/) for visualizing scientific and engineering data. Though I think that non-interactive plots with e.g. R/ggplot2 are more common for the diagnostics that I'm doing.

**Reply by lukego** [reply]
Relatedly: There was a company at ELS who employ ~15 data scientists and use an in-house Emacs-based user interface for their tools. They had recently hired a dedicated person to maintain their Emacs tooling full time. This seems quite similar to building tooling on kons-9.

**Reply by kaveh808** [reply]
I'd be curious to know more about the company at ELS and their use of Emacs for data visualization.

**Reply by lukego** [reply]
I don't know exactly what they do with Emacs but the company is [Machine Learning Programs](https://www.mlprograms.com/). My understanding is that they do machine learning models for the insurance industry. They are using Python on the backend. Breanndán Ó Nualláin is the person I talked with (he was the local organizer in Amsterdam for ELS this year.)

**Reply by kaveh808** [reply]
Cool, thanks.

**Comment by lukego** [comment]
My next diagnostic turned out to fit pretty well with heightmaps: https://github.com/lukego/blog/issues/39.
